File: src/main/java/com/ptmanager/backend/workplace/WorkplaceService.java
package com.ptmanager.backend.workplace;

import java.security.SecureRandom;
import java.util.List;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.ptmanager.backend.domain.Workplace;
import com.ptmanager.backend.repository.WorkplaceRepository;

@Service
public class WorkplaceService {

    private static final String CODE_ALPHABET = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789";
    private static final int CODE_LENGTH = 6;
    private static final SecureRandom RANDOM = new SecureRandom();

    private final WorkplaceRepository workplaceRepository;

    public WorkplaceService(WorkplaceRepository workplaceRepository) {
        this.workplaceRepository = workplaceRepository;
    }

    public List<Workplace> findWorkplaces() {
        return workplaceRepository.findAll();
    }

    @Transactional
    public Workplace createWorkplace(String name, String address) {
        Workplace workplace = new Workplace(null, name, address, generateUniqueInviteCode());
        return workplaceRepository.save(workplace);
    }

    private String generateUniqueInviteCode() {
        String code;
        do {
            StringBuilder builder = new StringBuilder(CODE_LENGTH);
            for (int i = 0; i < CODE_LENGTH; i++) {
                builder.append(CODE_ALPHABET.charAt(RANDOM.nextInt(CODE_ALPHABET.length())));
            }
            code = builder.toString();
        } while (workplaceRepository.existsByInviteCode(code));
        return code;
    }
}

File: src/main/kotlin/com/ptmanager/backend/workplace/WorkplaceService.kt
package com.ptmanager.backend.workplace

import com.ptmanager.backend.common.access.WorkplaceAccessGuard
import com.ptmanager.backend.domain.User
import com.ptmanager.backend.domain.UserRole
import com.ptmanager.backend.domain.Workplace
import com.ptmanager.backend.repository.UserRepository
import com.ptmanager.backend.repository.WorkplaceRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.security.SecureRandom
import java.util.NoSuchElementException

@Service
class WorkplaceService(
    private val workplaceRepository: WorkplaceRepository,
    private val userRepository: UserRepository,
    private val accessGuard: WorkplaceAccessGuard,
) {

    fun getWorkplace(id: Long): Workplace {
        accessGuard.requireMemberOf(id)
        return workplaceRepository.findById(id)
            .orElseThrow { NoSuchElementException("Workplace not found.") }
    }

    fun findMembers(workplaceId: Long, role: UserRole?): List<User> {
        accessGuard.requireMemberOf(workplaceId)
        return if (role == null) {
            userRepository.findByWorkplaceId(workplaceId)
        } else {
            userRepository.findByWorkplaceIdAndRole(workplaceId, role)
        }
    }

    @Transactional
    fun createWorkplace(name: String, address: String?): Workplace {
        val workplace = workplaceRepository.save(
            Workplace(name = name, address = address, inviteCode = generateUniqueInviteCode()),
        )
        // 생성자는 해당 매장에 소속된다.
        val creator = userRepository.findById(accessGuard.currentUserId())
            .orElseThrow { NoSuchElementException("User not found.") }
        creator.workplaceId = workplace.id
        userRepository.save(creator)
        return workplace
    }

    private fun generateUniqueInviteCode(): String {
        var code: String
        do {
            code = buildString {
                repeat(CODE_LENGTH) {
                    append(CODE_ALPHABET[RANDOM.nextInt(CODE_ALPHABET.length)])
                }
            }
        } while (workplaceRepository.existsByInviteCode(code))
        return code
    }

    companion object {
        private const val CODE_ALPHABET = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789"
        private const val CODE_LENGTH = 6
        private val RANDOM = SecureRandom()
    }
}
